cdap-watchdog/src/main/java/co/cask/cdap/logging/read/AvroFileReader.java

```diff
@@ -117,53 +118,44 @@ public void readLog(Location file, Filter logFilter, long fromTimeMs, long toTim
           return ImmutableList.of();
         }
 
-        GenericRecord datum;
         List<List<LogEvent>> logSegments = Lists.newArrayList();
         int count = 0;
 
         // Calculate skipLen based on fileLength
-        long skipLen = file.length() / 10;
+        long length = file.length();
+        LOG.trace("Got file Length {}", length);
+        long skipLen = length / 10;
         if (skipLen > DEFAULT_SKIP_LEN) {
           skipLen = DEFAULT_SKIP_LEN;
         } else if (skipLen <= 0) {
           skipLen = DEFAULT_SKIP_LEN;
         }
 
-        List<LogEvent> logSegment = Lists.newArrayList();
-
-        long lastSeekPos;
-        long seekPos = file.length();
-        while (seekPos > 0) {
-          lastSeekPos = seekPos;
-          seekPos = seekPos < skipLen ? 0 : seekPos - skipLen;
-          dataFileReader.sync(seekPos);
-
-          logSegment = logSegment.isEmpty() ? logSegment : Lists.<LogEvent>newArrayList();
-          // read all the elements in the current segment (seekPos up to lastSeekPos)
-          while (dataFileReader.hasNext() && !dataFileReader.pastSync(lastSeekPos)) {
-            datum = dataFileReader.next();
-
-            ILoggingEvent loggingEvent = LoggingEvent.decode(datum);
-
-            // Stop when reached fromTimeMs
-            if (loggingEvent.getTimeStamp() > fromTimeMs) {
-              break;
-            }
+        // For open file, start from file length and read till the actual eof
+        dataFileReader.sync(length);
+        long finalSync = dataFileReader.tell();
+        List<LogEvent> logSegment;
+        logSegment = readToNextSync(dataFileReader, logFilter, fromTimeMs, -1);
 
-            if (logFilter.match(loggingEvent)) {
-              ++count;
-              logSegment.add(new LogEvent(loggingEvent,
-                                          new LogOffset(LogOffset.INVALID_KAFKA_OFFSET, loggingEvent.getTimeStamp())));
-            }
-          }
+        long startPosition = finalSync;
+        long endPosition = startPosition;
+        long prevSync = finalSync;
 
-          if (!logSegment.isEmpty()) {
+        while (startPosition > 0 && count < maxEvents) {
+          if (logSegment != null && !logSegment.isEmpty()) {
             logSegments.add(logSegment);
+            count = count + logSegment.size();
```

Comment:
If `count` is incremented after the check `count < maxEvents` is made, then we'll be doing an extra `readToNextSync()` and discarding its results, right?